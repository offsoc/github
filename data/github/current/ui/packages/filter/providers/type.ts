import {fuzzyScore} from '@github-ui/fuzzy-score/fuzzy-score'
import {hasMatch} from 'fzy.js'
import type {Disposable, Environment} from 'react-relay'
import {fetchQuery, graphql} from 'react-relay'
import {createOperationDescriptor, getRequest} from 'relay-runtime'

import {FILTER_KEYS, FILTER_VALUES} from '../constants/filter-constants'
import type {FilterQuery} from '../filter-query'
import {
  type AnyBlock,
  type ARIAFilterSuggestion,
  type FilterConfig,
  type FilterProvider,
  FilterProviderType,
  type FilterValueData,
  FilterValueType,
  type IndexedBlockValueItem,
  type MutableFilterBlock,
  type SuppliedFilterProviderOptions,
  type ValueRowProps,
} from '../types'
import {
  caseInsensitiveStringCompare,
  escapeString,
  getFilterValue,
  getLastFilterBlockValue,
  getUnescapedFilterValue,
} from '../utils'
import {ValueIcon} from '../utils/ValueIcon'
import {AsyncFilterProvider} from '.'
import type {typeFilterIssueTypeQuery} from './__generated__/typeFilterIssueTypeQuery.graphql'
import type {typeFilterProjectIssueTypeQuery} from './__generated__/typeFilterProjectIssueTypeQuery.graphql'
import type {typeFilterViewerIssueTypeQuery} from './__generated__/typeFilterViewerIssueTypeQuery.graphql'

const RepoIssueTypesQuery = graphql`
  query typeFilterIssueTypeQuery($name: String!, $owner: String!) {
    repository(name: $name, owner: $owner) {
      issueTypes(first: 10) {
        edges {
          node {
            name
          }
        }
      }
    }
  }
`

const ViewerIssueTypesQuery = graphql`
  query typeFilterViewerIssueTypeQuery {
    viewer {
      suggestedIssueTypeNames
    }
  }
`

const ProjectIssueTypesQuery = graphql`
  query typeFilterProjectIssueTypeQuery($login: String!, $number: Int!) {
    organization(login: $login) {
      projectV2(number: $number) {
        suggestedIssueTypeNames
      }
    }
  }
`

type IssueType = {
  name: string
}

type RepositoryRequestVariables = {
  name: string
  owner: string
}

type ProjectRequestVariables = {
  login: string
  number: number
}

type ProjectScope = {
  login?: string
  projectNumber?: number
}

type RequestVariables = RepositoryRequestVariables | ProjectRequestVariables

export class TypeFilterProvider extends AsyncFilterProvider<IssueType> implements FilterProvider {
  filterValues?: ARIAFilterSuggestion[]
  relayEnvironment?: Environment
  repositoryScope?: string
  projectScope?: ProjectScope
  issueTypesEnabled?: boolean
  owner?: string
  repo?: string
  legacy: boolean
  requestVariables?: RequestVariables
  requestDisposable?: Disposable

  constructor(
    options?: SuppliedFilterProviderOptions,
    legacy: boolean = false,
    relayEnvironment?: Environment,
    repositoryScope?: string,
    issueTypesEnabled: boolean = false,
    projectScope?: ProjectScope,
  ) {
    super(FILTER_KEYS.type, {...options, filterTypes: {...options?.filterTypes, multiValue: true}})
    this.type = FilterProviderType.Text
    this.relayEnvironment = relayEnvironment
    this.repositoryScope = repositoryScope
    this.projectScope = projectScope
    this.issueTypesEnabled = issueTypesEnabled
    this.requestVariables = this.getRequestVariables()
    this.requestDisposable = this.getRequestDisposable()
    this.legacy = legacy
  }

  async getSuggestions(
    filterQuery: FilterQuery,
    filterBlock: AnyBlock | MutableFilterBlock,
    config: FilterConfig,
    caretIndex?: number | null,
  ) {
    const lastValue = getLastFilterBlockValue(filterBlock, caretIndex)
    const suggestions =
      (await this.processSuggestions(filterQuery, filterBlock, this.processSuggestion.bind(this), caretIndex)) || []
    const rankedSuggestions = []

    for (const suggestion of suggestions) {
      const {value: fValue, displayName} = suggestion
      const value = getFilterValue(fValue)

      const matchesValue = !!value && hasMatch(lastValue, value)
      const matchesDisplayName = !!displayName && hasMatch(lastValue, displayName)

      if (!value || (!matchesValue && !matchesDisplayName)) continue

      suggestion.priority -= fuzzyScore(lastValue, value)

      rankedSuggestions.push(suggestion)
    }

    return rankedSuggestions
  }

  override async fetchSuggestions(value: string, block: AnyBlock | MutableFilterBlock): Promise<IssueType[] | null> {
    if (!block) return []

    try {
      if (this.relayEnvironment && !!this.requestVariables && this.repositoryScope && this.issueTypesEnabled) {
        const repoIssueTypes = await fetchQuery<typeFilterIssueTypeQuery>(
          this.relayEnvironment,
          RepoIssueTypesQuery,
          this.requestVariables as RepositoryRequestVariables,
          {fetchPolicy: 'store-or-network'},
        ).toPromise()

        const edges = repoIssueTypes?.repository?.issueTypes?.edges || []
        const issueTypes = []
        for (const edge of edges) {
          if (edge && edge.node && edge.node.name) {
            issueTypes.push({name: edge.node.name})
          }
        }
        return issueTypes
      }
      // for memex we fetch the issue types from the memex_owner
      else if (this.relayEnvironment && this.projectScope && this.issueTypesEnabled) {
        const projectIssueTypes = await fetchQuery<typeFilterProjectIssueTypeQuery>(
          this.relayEnvironment,
          ProjectIssueTypesQuery,
          this.requestVariables as ProjectRequestVariables,
          {fetchPolicy: 'store-or-network'},
        ).toPromise()

        const issueTypes = projectIssueTypes?.organization?.projectV2?.suggestedIssueTypeNames || []
        return issueTypes?.map((item: string) => ({name: item})) as IssueType[]
      }

      // for /issues dashboard we need to fetch issue types from all orgs that the viewer is member of
      else if (this.relayEnvironment && !this.projectScope && !this.repositoryScope) {
        const viewerIssueTypes = await fetchQuery<typeFilterViewerIssueTypeQuery>(
          this.relayEnvironment,
          ViewerIssueTypesQuery,
          {},
          {fetchPolicy: 'store-or-network'},
        ).toPromise()

        const issueTypes = viewerIssueTypes?.viewer?.suggestedIssueTypeNames || []
        return issueTypes?.map((item: string) => ({name: item})) as IssueType[]
      }
    } catch {
      return []
    }

    return []
  }

  override async validateFilterValue(value: string): Promise<IssueType | null> {
    try {
      if (this.relayEnvironment && !!this.requestVariables && this.repositoryScope && this.issueTypesEnabled) {
        const repoIssueTypes = await fetchQuery<typeFilterIssueTypeQuery>(
          this.relayEnvironment,
          RepoIssueTypesQuery,
          this.requestVariables as RepositoryRequestVariables,
          {fetchPolicy: 'store-or-network'},
        ).toPromise()

        const edges = repoIssueTypes?.repository?.issueTypes?.edges || []
        const issueTypes = []
        for (const edge of edges) {
          if (edge && edge.node && edge.node.name) {
            issueTypes.push({name: edge.node.name})
          }
        }
        return issueTypes.find(i => caseInsensitiveStringCompare(i.name, value)) as IssueType
      } else if (this.relayEnvironment && !!this.requestVariables && this.projectScope && this.issueTypesEnabled) {
        const projectIssueTypes = await fetchQuery<typeFilterProjectIssueTypeQuery>(
          this.relayEnvironment,
          ProjectIssueTypesQuery,
          this.requestVariables as ProjectRequestVariables,
          {fetchPolicy: 'store-or-network'},
        ).toPromise()

        const issueTypes = projectIssueTypes?.organization?.projectV2?.suggestedIssueTypeNames || []
        return {name: issueTypes?.find(item => caseInsensitiveStringCompare(item, value))} as IssueType
      } else if (this.relayEnvironment) {
        const viewerIssueTypes = await fetchQuery<typeFilterViewerIssueTypeQuery>(
          this.relayEnvironment,
          ViewerIssueTypesQuery,
          {},
          {fetchPolicy: 'store-or-network'},
        ).toPromise()

        const issueTypes = viewerIssueTypes?.viewer?.suggestedIssueTypeNames || []
        return {name: issueTypes?.find(item => caseInsensitiveStringCompare(item, value))} as IssueType
      }
    } catch {
      return null
    }
    return null
  }

  override validateValue(
    filterQuery: FilterQuery,
    value: IndexedBlockValueItem,
    issueType: IssueType | null,
  ): false | Partial<IndexedBlockValueItem> | null | undefined {
    const extractedValue = getUnescapedFilterValue(value.value)
    const validValues = this.legacy
      ? [...FILTER_VALUES.type.map(({value: typeValue}) => typeValue as string), issueType?.name]
      : [issueType?.name]
    if (validValues.length === 0 || !extractedValue) return false

    const matchedName = validValues.find(name => caseInsensitiveStringCompare(name, extractedValue))
    if (!matchedName) return false

    return {
      value: escapeString(extractedValue) ?? '',
      displayName: extractedValue,
    }
  }

  override async processSuggestions(
    filterQuery: FilterQuery,
    filterBlock: AnyBlock | MutableFilterBlock,
    processSuggestion: (item: IssueType, query: string, filterQuery: FilterQuery) => ARIAFilterSuggestion,
    caretIndex?: number | null,
  ) {
    if (!this.shouldGetSuggestions(filterBlock)) return null

    const lastValue = getLastFilterBlockValue(filterBlock, caretIndex)
    const items = (await this.fetchSuggestions(lastValue, filterBlock)) ?? []
    const suggestions: ARIAFilterSuggestion[] = this.legacy
      ? FILTER_VALUES.type.map(fv => ({...fv, ariaLabel: `${fv.displayName}, ${this.displayName}`}))
      : []

    return [...suggestions, ...items.map(item => processSuggestion(item, lastValue, filterQuery))]
  }

  getValueRowProps(value: FilterValueData): ValueRowProps {
    return {
      text: value.displayName ?? getFilterValue(value.value) ?? '',
      leadingVisual: ValueIcon({value, providerIcon: this.icon}),
    }
  }

  private processSuggestion(issueType: IssueType, query: string): ARIAFilterSuggestion {
    const {name} = issueType
    let priority = 3

    if (query) {
      if (name) priority -= fuzzyScore(query, name)
    }

    return {
      type: FilterValueType.Value,
      displayName: name,
      ariaLabel: `${name}, ${this.displayName}`,
      value: escapeString(name) ?? '',
      inlineDescription: true,
      priority,
      icon: this.icon,
    }
  }

  private getRequestVariables(): RequestVariables | undefined {
    if (this.repositoryScope) {
      const repoParts = this.repositoryScope.split('/')

      return {
        owner: repoParts[0] || '',
        name: repoParts[1] || '',
      }
    } else if (this.projectScope?.login && this.projectScope?.projectNumber) {
      return {
        login: this.projectScope.login,
        number: this.projectScope.projectNumber,
      }
    } else {
      return undefined
    }
  }

  private getRequestDisposable(): Disposable | undefined {
    if (!this.relayEnvironment) {
      return undefined
    }

    if (this.requestVariables && this.repositoryScope) {
      const request = getRequest(RepoIssueTypesQuery)
      const operation = createOperationDescriptor(request, this.requestVariables)
      return this.relayEnvironment.retain(operation)
    } else if (this.requestVariables && this.projectScope) {
      const request = getRequest(ProjectIssueTypesQuery)
      const operation = createOperationDescriptor(request, this.requestVariables)
      return this.relayEnvironment.retain(operation)
    } else {
      const request = getRequest(ViewerIssueTypesQuery)
      const operation = createOperationDescriptor(request, {})
      return this.relayEnvironment.retain(operation)
    }
  }
}
