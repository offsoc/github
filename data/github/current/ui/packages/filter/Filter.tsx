import {type TestIdProps, testIdProps} from '@github-ui/test-id-props'
import {Box, type SxProp, useResponsiveValue} from '@primer/react'
import type {ResponsiveValue} from '@primer/react/lib-esm/hooks/useResponsiveValue'
import type React from 'react'
import {useCallback, useState} from 'react'

import {AdvancedFilterDialog, type FilterButtonVariant} from './advanced-filter-dialog/AdvancedFilterDialog'
import {ClearButton} from './clear-button/ClearButton'
import {defaultFilterSettings} from './constants/defaults'
import {FilterContextProvider} from './context/RootContext'
import {FilterQuery} from './filter-query'
import {FilterInputIcon} from './FilterInputIcon'
import {FilterInputWrapper} from './FilterInputWrapper'
import {Input} from './input/Input'
import {SubmitButton} from './input/SubmitButton'
import {SuggestionsList} from './suggestions/SuggestionsList'
import type {FilterProvider, FilterSettings, FilterVariant, SubmitEvent} from './types'
import {ValidationMessage} from './ValidationMessage'

// Exports from the root of the package
export * from './constants/defaults'
export * from './filter-query'
export * from './FilterRevert'
export * from './types'
export * from './ValidationMessage'

const filterInputWrapperStyles = ({variant, onSubmit}: {variant: FilterProps['variant']; onSubmit: boolean}) => {
  const leftRounded = variant === 'input'
  const rightRounded = (variant === 'input' && !onSubmit) || (variant !== 'input' && !onSubmit)

  return {
    display: 'grid',
    gridTemplateColumns: onSubmit ? '1fr min-content' : 'min-content 1fr min-content',
    alignItems: 'center',
    pl: onSubmit ? '12px' : 1,
    pr: '1px',
    width: '100%',
    borderStyle: 'solid',
    borderWidth: '1px',
    borderColor: 'border.default',
    marginInlineEnd: -1,
    ...(leftRounded && {
      borderTopLeftRadius: 2,
      borderBottomLeftRadius: 2,
    }),
    ...(rightRounded && {
      borderTopRightRadius: 2,
      borderBottomRightRadius: 2,
    }),
    ':has(input:focus-visible)': {
      zIndex: 1,
      outline: '2px solid var(--borderColor-accent-emphasis, var(--color-accent-emphasis))',
      outlineOffset: '-2px',
    },
  }
}

export type FilterProps = {
  /** ID of the Filter Component */
  id: string
  /** Change event callback, passing the updated filter value */
  onChange?: (value: string) => void
  /** Event callback, triggered by the filter query being parsed, and passing the updated filter query request*/
  onParse?: (request: FilterQuery) => void
  /** Submit event callback, triggered by Enter or pressing Submit, and passing the updated filter query and the
   * originating event type */
  onSubmit?: (request: FilterQuery, eventType: SubmitEvent) => void
  /** Array of FilterProviders that are being used/supported */
  providers: FilterProvider[]
  /** Context to be applied for all API queries */
  context?: Record<string, string>
  /** Variant of the filter button, whether to show the title or just the icon */
  filterButtonVariant?: FilterButtonVariant
  /** Initial value to use for the input. DO NOT USE in conjuction with `filterValue`.
   * Only use this if you want the Filter to have an initial value but remain uncontrolled. */
  initialFilterValue?: string
  /** Supplied filter value to use. By using this, the Filter is externally controlled
   * and will not use it's own state. */
  filterValue?: string
  /** React Ref to the input element */
  inputRef?: React.RefObject<HTMLInputElement>
  /** Label displayed atop the Filter bar. This is only shown when `visuallyHideLabel` is `false` */
  label: string
  /** Input keydown event callback */
  onKeyDown?: React.KeyboardEventHandler<Element>
  /** Validation callback, passing the validation message and the validation query */
  onValidation?: (messages: string[], filterQuery: FilterQuery) => void
  /** Placeholder text for the input */
  placeholder?: string
  /** Settings for the Filter, such as support of the legacy No filter provider and alias matching */
  settings?: FilterSettings
  /** Whether to show any validation messages or not. If this is disabled, use `onValidation` to receive the messages to display elsewhere */
  showValidationMessage?: boolean
  /** Variant of the Filter to render. `button` or `input` for one or the other, `full` (default) renders both.  */
  variant?: FilterVariant
  /** Whether to visually hide the Label or not */
  visuallyHideLabel?: boolean
} & TestIdProps &
  SxProp

const defaultFilterQuery = new FilterQuery()

export function Filter({
  context,
  'data-testid': dataTestId,
  filterButtonVariant = 'normal',
  filterValue,
  id,
  initialFilterValue,
  inputRef,
  label,
  onChange,
  onParse,
  onSubmit,
  onKeyDown,
  onValidation,
  placeholder,
  providers,
  settings,
  showValidationMessage = true,
  sx,
  variant = 'full',
  visuallyHideLabel = true,
}: FilterProps) {
  const [validationMessage, setValidationMessage] = useState<string[]>([])
  const hasValidationMessage = showValidationMessage && validationMessage.length > 0

  const validationCallback = useCallback(
    (messages: string[] = [], filterQuery: FilterQuery = defaultFilterQuery) => {
      setValidationMessage(messages)
      onValidation?.(messages, filterQuery)
    },
    [onValidation],
  )

  const filterButtonResponsive = useResponsiveValue<ResponsiveValue<FilterButtonVariant>, FilterButtonVariant>(
    {regular: filterButtonVariant, narrow: variant === 'button' ? 'normal' : 'compact'},
    filterButtonVariant,
  )

  const externallyControlled = typeof filterValue === 'string'
  const [innerValue, setInnerValue] = useState<string>(initialFilterValue || '')

  const changeCallback = useCallback(
    (value: string) => {
      if (!externallyControlled) {
        setInnerValue(value)
      }
      onChange?.(value)
    },
    [externallyControlled, onChange],
  )

  return (
    <FilterContextProvider
      context={context}
      inputRef={inputRef}
      rawFilter={externallyControlled ? filterValue : innerValue}
      onChange={changeCallback}
      onParse={onParse}
      onSubmit={onSubmit}
      onValidation={validationCallback}
      providers={providers}
      settings={{...defaultFilterSettings, ...settings}}
      variant={variant}
    >
      <Box
        id={id}
        role="form"
        className="FormControl FormControl--fullWidth"
        sx={{...sx, gap: 0}}
        {...testIdProps(dataTestId ?? 'filter')}
      >
        <label
          id={`${id}-label`}
          htmlFor={`${id}-input`}
          className={`FormControl-label ${visuallyHideLabel ? 'sr-only' : ''}`}
          {...testIdProps('filter-bar-label')}
        >
          {label}
        </label>
        <Box sx={{display: 'grid', gap: 3, width: '100%', position: 'relative'}}>
          <FilterInputWrapper isStandalone={variant === 'button'}>
            {variant !== 'input' && (
              <AdvancedFilterDialog isStandalone={variant === 'button'} filterButtonVariant={filterButtonResponsive} />
            )}
            {variant !== 'button' && (
              <>
                <Box
                  sx={filterInputWrapperStyles({
                    variant,
                    onSubmit: !!onSubmit,
                  })}
                >
                  {!onSubmit && <FilterInputIcon />}
                  <Input
                    id={id}
                    placeholder={placeholder}
                    onKeyDown={onKeyDown}
                    hasValidationMessage={hasValidationMessage}
                  />
                  <ClearButton />
                </Box>
                {onSubmit && <SubmitButton />}
              </>
            )}
          </FilterInputWrapper>
          <SuggestionsList id={id} />
          {/* This needs to move outside the Filter component because we can't predict where the consumer wants to put the error message. In memex this would go above the table view. */}
          {showValidationMessage && <ValidationMessage messages={validationMessage} id={`${id}-validation-message`} />}
        </Box>
      </Box>
    </FilterContextProvider>
  )
}
