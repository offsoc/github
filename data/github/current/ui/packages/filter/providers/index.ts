export {AsyncFilterProvider} from './async'
export * from './key-only'
export {LabelFilterProvider} from './label'
export {LanguageFilterProvider} from './language'
export {MilestoneFilterProvider} from './milestone'
export {OrgFilterProvider} from './org'
export {ProjectFilterProvider} from './project'
export {RepositoryFilterProvider} from './repository'
export * from './static'
export {TeamFilterProvider, TeamReviewRequestedFilterProvider} from './team'
export {TypeFilterProvider} from './type'
export * from './user'
