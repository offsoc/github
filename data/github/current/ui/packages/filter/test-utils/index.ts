import {act, fireEvent, screen} from '@testing-library/react'

export const setupExpectedAsyncErrorHandler = () => {
  jest.spyOn(console, 'error').mockImplementation((message: string) => {
    // * Because Filter is asynchronous, there are console errors that are thrown, but expected. This will rethrow
    // * any errors that are not related to the async nature of the component.
    if (!message.includes?.('wrapped in act(')) {
      // eslint-disable-next-line no-console
      console.warn(message)
    }
  })
}

export async function updateFilterValue(key: string, value: string = '') {
  jest.useFakeTimers()

  const input = screen.getByRole('combobox')
  input.focus()

  fireEvent.change(input, {target: {value: `${key}${key && value ? ':' : ''}${value}`}})

  // eslint-disable-next-line @typescript-eslint/require-await
  await act(async () => {
    jest.runAllTimers()
  })
}
