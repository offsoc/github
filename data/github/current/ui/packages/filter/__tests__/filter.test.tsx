import {render} from '@github-ui/react-core/test-utils'
import {screen, waitFor} from '@testing-library/react'
import {useState} from 'react'

import type {FilterProps} from '../Filter'
import {Filter} from '../Filter'
import {StateFilterProvider, StatusFilterProvider} from '../providers'
import {updateFilterValue} from '../test-utils'
import {SubmitEvent} from '../types'
import {
  appendToFilterAndRenderAsyncSuggestions,
  expectFilterValueToBe,
  expectSuggestionsToMatchSnapshot,
  selectSuggestion,
  setupAsyncErrorHandler,
} from './utils/helpers'

describe('Filter', () => {
  setupAsyncErrorHandler()

  it('renders both a filter input and a filter button by default', () => {
    render(<Filter id="test-filter-bar" label="Filter" providers={[]} />)
    expect(screen.getByRole('combobox')).toBeInTheDocument()
    expect(screen.getByRole('button')).toBeInTheDocument()
  })
  it('renders only a filter input if the variant is set to input', () => {
    render(<Filter id="test-filter-bar" label="Filter" providers={[]} variant="input" />)
    expect(screen.getByRole('combobox')).toBeInTheDocument()
    expect(screen.queryByRole('button')).not.toBeInTheDocument()
  })
  it('renders only a filter button if the variant is set to button', () => {
    render(<Filter id="test-filter-bar" label="Filter" providers={[]} variant="button" />)
    expect(screen.queryByRole('combobox')).not.toBeInTheDocument()
    expect(screen.getByRole('button')).toBeInTheDocument()
  })
  it('renders a normal button by default', () => {
    render(<Filter id="test-filter-bar" label="Filter" providers={[]} />)
    expect(screen.getByRole('button')).toHaveTextContent('Filter')
  })
  it('renders a compact button if the filterButtonVariant is set to compact', () => {
    render(<Filter id="test-filter-bar" label="Filter" providers={[]} filterButtonVariant="compact" />)
    expect(screen.getByRole('button')).not.toContain('Filter')
  })

  it('renders initialFilterValue', async () => {
    const initialFilterValue = 'initial-value'

    render(<Filter id="test-filter-bar" label="Filter" providers={[]} initialFilterValue={initialFilterValue} />)

    expect(await screen.findByTestId('styled-input-content')).toHaveTextContent(initialFilterValue)
    expectFilterValueToBe(initialFilterValue)
  })

  it('should render a screen-reader only label by default', () => {
    render(<Filter id="test-filter-bar" label="Filter" providers={[]} />)
    expect(screen.getByTestId('filter-bar-label').classList).toContain('sr-only')
  })
  it('should render a visible label', () => {
    render(<Filter id="test-filter-bar" label="Filter" visuallyHideLabel={false} providers={[]} />)
    expect(screen.getByTestId('filter-bar-label').classList).not.toContain('sr-only')
  })

  it('expects aria-controls to match ID of suggestions list', async () => {
    const filterProviders = [new StateFilterProvider('mixed')]

    render(<Filter id="test-filter-bar" label="Filter" providers={filterProviders} />)
    await updateFilterValue('s')
    expect(screen.getByRole('combobox')).toHaveAttribute('aria-controls', 'test-filter-bar-results')
    expect(screen.getByRole('listbox')).toHaveAttribute('id', 'test-filter-bar-results')
  })

  it('expects aria-activedescendant to be undefined when no suggestion is active', async () => {
    const filterProviders = [new StateFilterProvider('mixed')]

    const {user} = render(<Filter id="test-filter-bar" label="Filter" providers={filterProviders} />)

    const input = screen.getByRole('combobox')
    await updateFilterValue('s')

    expect(input).not.toHaveAttribute('aria-activedescendant')
    await user.keyboard('{ArrowDown}')
    expect(input).toHaveAttribute('aria-activedescendant', 'suggestion-0')
  })

  it('suggestion item should only have aria-label defined with filter type', async () => {
    const filterProviders = [new StateFilterProvider('mixed')]

    render(<Filter id="test-filter-bar" label="Filter" providers={filterProviders} />)
    await updateFilterValue('')

    expect(screen.getAllByRole('option')[0]).toHaveAttribute(
      'aria-label',
      'State, Filter, Look for open / closed items',
    )
    expect(screen.getAllByRole('option')[0]).not.toHaveAttribute('aria-labelledby')
  })

  describe('Events', () => {
    it('calls the onChange callback when the filter input changes', async () => {
      const onChange = jest.fn()
      const {user} = render(<Filter id="test-filter-bar" label="Filter" providers={[]} onChange={onChange} />)

      const input = screen.getByRole('combobox')
      await user.type(input, '1234')

      await waitFor(() => {
        expect(onChange).toHaveBeenCalledTimes(4)
      })
    })

    it('calls the onKeyDown callback when a key is pressed in the filter input', async () => {
      const onKeyDown = jest.fn()
      const {user} = render(<Filter id="test-filter-bar" label="Filter" providers={[]} onKeyDown={onKeyDown} />)
      const input = screen.getByRole('combobox')
      await user.type(input, '{Enter}')

      await waitFor(() => {
        expect(onKeyDown).toHaveBeenCalled()
      })
    })

    it('calls the onValidation callback when the filter input is validated', async () => {
      const onValidation = jest.fn()
      render(
        <Filter
          id="test-filter-bar"
          label="Filter"
          providers={[]}
          onValidation={onValidation}
          showValidationMessage={true}
        />,
      )

      await updateFilterValue('1234')

      await waitFor(() => {
        expect(onValidation).toHaveBeenCalled()
      })
    })

    it('calls onValidation with filter query', async () => {
      const onValidation = jest.fn()
      render(
        <Filter
          id="test-filter-bar"
          label="Filter"
          providers={[]}
          onValidation={onValidation}
          showValidationMessage={true}
        />,
      )

      await updateFilterValue('1234')

      await waitFor(() => {
        expect(onValidation).toHaveBeenLastCalledWith([], expect.objectContaining({raw: '1234'}))
      })
    })

    it('does not highlight if query is invalid (unbalanced quotes)', async () => {
      const onValidation = jest.fn()
      render(<Filter id="test-filter-bar" label="Filter" providers={[]} onValidation={onValidation} />)
      const defaultFilterQueryObject = expect.objectContaining({raw: ''})

      await updateFilterValue('is:"')

      await waitFor(() => {
        // When quotes are not balanced, error blocks are not processed
        expect(onValidation).toHaveBeenCalledWith([], defaultFilterQueryObject)
      })

      await appendToFilterAndRenderAsyncSuggestions('issue"')

      await waitFor(() => {
        expect(onValidation).toHaveBeenCalledWith([], defaultFilterQueryObject)
      })
    })

    it('should trigger an external form submission on "Enter" when a suggestion is not selected or active', async () => {
      const onSubmit = jest.fn()

      const {user} = render(
        <form id="test" onSubmit={onSubmit}>
          <Filter id="test-filter-bar" label="Filter" providers={[]} />
        </form>,
      )
      const input = screen.getByRole('combobox')

      // Deselect the active suggestion
      await user.type(input, '{home}')

      // Trigger form submission
      await user.type(input, '{enter}')

      await waitFor(() => {
        expect(onSubmit).toHaveBeenCalled()
      })
    })

    it('should not trigger an external form submission on "Enter" when a filter suggestion is selected or active', async () => {
      const onSubmit = jest.fn()

      const {user} = render(
        <form id="test" onSubmit={onSubmit}>
          <Filter id="test-filter-bar" label="Filter" providers={[]} />
        </form>,
      )
      const input = screen.getByRole('combobox')

      // Trigger form submission
      await user.type(input, '{enter}')

      await waitFor(() => {
        expect(onSubmit).toHaveBeenCalled()
      })
    })

    it('should trigger the onSubmit callback on "Enter" when a filter value suggestion is selected', async () => {
      const onSubmit = jest.fn()

      const filterProviders = [new StateFilterProvider(), new StatusFilterProvider()]
      render(<Filter id="test-filter-bar" label="Filter" providers={filterProviders} onSubmit={onSubmit} />)
      const input = screen.getByRole('combobox')

      await waitFor(() => {
        expect(onSubmit).not.toHaveBeenCalled()
      })

      input.focus()

      await updateFilterValue('st')

      expectSuggestionsToMatchSnapshot()
      await selectSuggestion('State')

      expectFilterValueToBe('state:')

      await waitFor(() => {
        expect(onSubmit).not.toHaveBeenCalled()
      })

      expectSuggestionsToMatchSnapshot()
      await selectSuggestion('Open')

      await waitFor(() => {
        expect(onSubmit).toHaveBeenCalled()
      })
      expect(onSubmit.mock.lastCall[1]).toBe(SubmitEvent.SuggestionSelected)
    })

    it('should trigger the onSubmit callback on "Enter" when a filter value suggestion is not selected', async () => {
      const onSubmit = jest.fn()

      const filterProviders = [new StateFilterProvider(), new StatusFilterProvider()]
      const {user} = render(
        <Filter id="test-filter-bar" label="Filter" providers={filterProviders} onSubmit={onSubmit} />,
      )
      const input = screen.getByRole('combobox')

      await waitFor(() => {
        expect(onSubmit).not.toHaveBeenCalled()
      })

      input.focus()

      await updateFilterValue('st')

      expectSuggestionsToMatchSnapshot()
      await selectSuggestion('State')

      expectFilterValueToBe('state:')

      await waitFor(() => {
        expect(onSubmit).not.toHaveBeenCalled()
      })

      await user.type(input, '{Enter}')

      await waitFor(() => {
        expect(onSubmit).toHaveBeenCalled()
      })
      expect(onSubmit.mock.lastCall[1]).toBe(SubmitEvent.ExplicitSubmit)
    })
  })

  describe('Externally controlled', () => {
    it('updates the input immediately', async () => {
      const onKeyDown = jest.fn()
      const {user} = render(<ExternallyControlledFilter onKeyDown={onKeyDown} />)
      const input = screen.getByRole('combobox')
      await user.type(input, '1256')
      expect(input).toHaveValue('1256')

      await user.keyboard('{ArrowLeft}{ArrowLeft}34')
      expect(input).toHaveValue('123456')

      await user.tripleClick(input)
      await user.keyboard('{Backspace}')
      expect(input).toHaveValue('')

      // This waitFor is necessary to avoid the async act() error
      await waitFor(() => expect(onKeyDown).toHaveBeenCalled())
    })
  })
})

function ExternallyControlledFilter(props: Partial<FilterProps>) {
  const [value, setValue] = useState('')
  return <Filter id="test-id" label="Filter!!!" providers={[]} filterValue={value} onChange={setValue} {...props} />
}

describe('Filter Input', () => {
  setupAsyncErrorHandler()

  it('renders a blank filter input if no value is provided', () => {
    render(<Filter id="test-filter-bar" label="Filter" providers={[]} />)
    expectFilterValueToBe('')
  })
  it('renders a filter input with the provided value', () => {
    render(<Filter id="test-filter-bar" label="Filter" providers={[]} filterValue="test" />)
    expectFilterValueToBe('test')
  })
  it('renders a filter input with the provided placeholder', () => {
    render(<Filter id="test-filter-bar" label="Filter" providers={[]} placeholder="test" />)
    expect(screen.getByRole('combobox')).toHaveAttribute('placeholder', 'test')
  })
})
