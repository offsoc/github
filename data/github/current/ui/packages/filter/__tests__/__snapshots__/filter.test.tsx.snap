// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Events should trigger the onSubmit callback on "Enter" when a filter value suggestion is not selected 1`] = `
[
  "State",
  "Status",
]
`;

exports[`Filter Events should trigger the onSubmit callback on "Enter" when a filter value suggestion is selected 1`] = `
[
  "State",
  "Status",
]
`;

exports[`Filter Events should trigger the onSubmit callback on "Enter" when a filter value suggestion is selected 2`] = `
[
  "Open",
  "Closed",
  "Merged",
  "Draft",
]
`;
